[DO NOT REVIEW] Channels last perf test

VitalyFedyunin opened the pull request:
This PR have combined changes of #23403 subtasks.

jjsjann123 committed:
[nhwc support]

Added cudnn nhwc support for:
1. batch norm
2. convolution
3. convolution_transpose

VitalyFedyunin committed:
Simplify tests that should cover all possible devices

VitalyFedyunin committed:
More examples

VitalyFedyunin committed:
Better messages

VitalyFedyunin committed:
empty_like, to and clone now preserve memory format if necessary

jjsjann123 committed:
patching suggest_memory_format for convolution

suggest_memory_format has ambiguous meaning for two cases:
1. tensor with NCHW where C = 1.
   we could use stride of C as a hint to tell the intended memory format.
2. tensor with NCHW where H == W == 1.
   there's no way to identify the intended memory format from strides.

Currently we fallback to NCHW whenever we see contiguous tensor. Hence avoiding
ambiguity for some of the special cases.

jjsjann123 committed:
nhwc test cases added for batchnorm & conv

jjsjann123 committed:
code cleaning and comments

VitalyFedyunin committed:
WIP

VitalyFedyunin committed:
resize_as_ memory format preservation

jjsjann123 committed:
addressing review comments

VitalyFedyunin committed:
Add Tensor Iterator and some cuda functions memory propagation

jjsjann123 committed:
removing obsolete change

jjsjann123 committed:
disable nhwc tests for rocm

jjsjann123 committed:
[Fixing is_channels_last_contiguous_ & is_channels_last_ flag]

The old implementation assumed `is_channels_last_contiguous_` to be mutually
exclusive to `is_contiguous_`, which is not true.
Properly set the flag by checking strides.

jjsjann123 committed:
test case added

VitalyFedyunin committed:
scatter and gather now support channels last

jjsjann123 committed:
removing tabs

jjsjann123 committed:
fixing pylint issue

jjsjann123 committed:
[nhwc support for adaptive_avg_pool2d & adaptive_avg_pool2d_backward]

Initial kernel support added for optimized NHWC tensor.

TODO: currently backwards kernel spits out tensor with NHWC stride.
Unfortunately autograd restores grad to contiguous (in either copy or add). This
makes real perf tuning annoying to do. (since I cannot easily measure end-to-end
time in my python script)

My current kernel is blazing fast comparing to the original NCHW kernel in fp16,
since I avoided atomicAdd. I'll finish perf tuning after we merged some future
PR expanding NHWC support in the core.

VitalyFedyunin committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into clone_empty_supports_memory_format

VitalyFedyunin committed:
Fix tests

ifedan committed:
Channel last memory format

ifedan committed:
Channel last memory format

ifedan committed:
Added tests

ifedan committed:
Channel last memory format

ifedan committed:
Channel last memory format

ifedan committed:
Channel last memory format

ifedan committed:
Revert "Channel last memory format"

This reverts commit c7ece816bc4df1b71449d14faa426b925cf516dc.

ifedan committed:
Channel last memory format

jjsjann123 committed:
perf tuning finished

jjsjann123 committed:
nhwc unit test added in test_nn.py

ifedan committed:
Channel last memory format. Multiple kernels

ifedan committed:
Channel last memory format. Multiple kernels

ifedan committed:
Performance improvement

ifedan committed:
Performance improvement

ifedan committed:
Changes based on PR

ifedan committed:
Merge branch 'master' into max_pool2d_cuda_perf

ifedan committed:
Changes based on PR review

ifedan committed:
Merge branch 'max_pool2d_cuda_perf' of https://github.com/ifedan/pytorch into max_pool2d_cuda_perf

ifedan committed:
Changes based on PR comments

VitalyFedyunin committed:
Merge branch 'pr_23899' into channels_last_perf_test

VitalyFedyunin committed:
Merge branch 'pr_24038' into channels_last_perf_test

VitalyFedyunin committed:
Merge branch 'pr_24121' into channels_last_perf_test

VitalyFedyunin committed:
Merge branch 'pr_24396' into channels_last_perf_test

VitalyFedyunin committed:
Merge branch 'pr_24872' into channels_last_perf_test

VitalyFedyunin committed:
Merge branch 'pr_23861' into channels_last_perf_test

VitalyFedyunin committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into channels_last_perf_test

VitalyFedyunin commented:
@jjsjann123  @ifedan FYI

VitalyFedyunin committed:
Fixing merge problems

VitalyFedyunin committed:
Merge branch 'pr_24113' into channels_last_perf_test

VitalyFedyunin committed:
WIP

jjsjann123 commented:
BTW, https://github.com/pytorch/pytorch/blob/fc7f4e4475a4a1e39446d3c8b3ffa55cb68ba64b/torch/csrc/autograd/functions/accumulate_grad.cpp#L42-L59

This guy clones the grad if it's not contiguous. In the benchmarking for my PR, that permutation kills the perf gain from NHWC :/ 

Not sure how it works out now as our `clone()` supports NHWC, but we may want to double check the kernels in a toy example to make sure there's no wasted permutation.

VitalyFedyunin committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into channels_last_perf_test

VitalyFedyunin committed:
WIP

VitalyFedyunin committed:
WIP

jjsjann123 committed:
[Adaptive_avg_pooling nhwc kernel fixes]

Previous kernel does not stride on Channel dimension, and the kernel uses shared
memory to store temporary result (to break data dependency -> code paralellism)

Resulted in requesting more resources than what's available.

Fixing:
added striding on C to reduce shmem usage per CTA.

VitalyFedyunin committed:
Remove unnecessary .contiguous as operator does it internally

VitalyFedyunin committed:
Debug lines

VitalyFedyunin committed:
Debug lines

VitalyFedyunin committed:
Support Channels Last

jjsjann123 committed:
[BatchNormalization cudnn API update]

Updated cudnn API for batchnorm. Enabling the Extended API which provides
semi-persistent batchnorm kernel that has better performance on NHWC layout.

TODO: I made adjustments to the API as well as BN in JIT IR. But I haven't fully
tested the JIT part yet. I should verify that in the final PR.
